lib/mayaUsd/ufe/StagesSubject.cpp

```diff
@@ -277,9 +265,29 @@ void StagesSubject::stageChanged(UsdNotice::ObjectsChanged const& notice, UsdSta
 	}
 }
 
-void StagesSubject::onStageSet(const UsdMayaProxyStageSetNotice& notice)
+void StagesSubject::onStageSet(const MayaUsdProxyStageSetNotice& notice)
+{
+	// We should have no listerners and stage map is dirty.
+	TF_VERIFY(g_StageMap.isDirty());
+	TF_VERIFY(fStageListeners.empty());
+
+	StagesSubject::Ptr me(this);
+	for (auto stage : ProxyShapeHandler::getAllStages())
+	{
+		fStageListeners[stage] = TfNotice::Register(
+			me, &StagesSubject::stageChanged, stage);
+	}
+}
+
+void StagesSubject::onStageInvalidate(const MayaUsdProxyStageInvalidateNotice& notice)
 {
 	afterOpen();
+
+#if UFE_PREVIEW_VERSION_NUM >= 2014
+	Ufe::SceneItem::Ptr sceneItem = Ufe::Hierarchy::createItem(notice.GetProxyShape().ufePath());
+	auto notification = Ufe::SubtreeInvalidate(sceneItem);
+	Ufe::Scene::notifySubtreeInvalidate(notification);
+#endif
```

Comment:
Send a new UFE subtree invalidate notice when the stage is invalidated. Clients like the Outliner can use this to re-build.